### PR TITLE
Associate reason with EventsDisallowed mode.

### PR DESCRIPTION
**Pull request by kannanvijayan**
This changes the signature of AutoDisallowEvents and its uses to associate a string reason.  That reason is currently stored in a global, but we should figure out how to pull it out and dump it into crash reports so we have easy access to it.

Not quite sure how to do that yet.

### CODE REVIEW

**Review by bhackett1024 — CHANGES_REQUESTED** [review]
This is a good idea, but we'll want this information in more places than just crash reports, e.g. warnings due to events being disallowed etc.  I think we should modify the RecordReplayBeginDisallowEvents API to take a reason string, and then the recorder can keep track of this internally for all the reporting needs it has.  This will require modifying all our applications which use this API, an alternative would be to add a variant API that takes the reason argument but that adds clutter.

**Comment by kannanvijayan** [comment]
@bhackett1024 what if we add a new hook to allow the runtime to register a function that lets the driver/linker pull the current disallow-events reason string when it needs?

e.g.
`RecordReplayRegisterEventsDisallowedReasonGetter(const char* (*getter)());`

**Comment by bhackett1024** [comment]
> @bhackett1024 what if we add a new hook to allow the runtime to register a function that lets the driver/linker pull the current disallow-events reason string when it needs?
> 
> e.g. `RecordReplayRegisterEventsDisallowedReasonGetter(const char* (*getter)());`

This seems a little circuitous, why don't we add a RecordReplayBeginDisallowEventsWithName API?  Looking through the API we actually did something like this recently, and have both RecordReplayRegisterPointer and RecordReplayRegisterPointerWithName APIs.

File: base/record_replay.h [hunk 1]
```diff
@@ -72,6 +73,7 @@ struct AutoPassThroughEvents {
 
 struct AutoDisallowEvents {
   AutoDisallowEvents() { BeginDisallowEvents(); }
```
Comment (bhackett1024) [inline base/record_replay.h:75]: Can we remove this constructor?  It seems like we'll always want to have a label.

**Review by bhackett1024 — APPROVED** [review]

**Pull request merged**